File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/A._Лампочки/solution.py
import os

LOCAL = os.environ.get('REMOTE_JUDGE', 'false') != 'true'

if LOCAL:
    class Node:
        def __init__(self, value, left=None, right=None):
            self.value = value
            self.right = right
            self.left = left


def solution(root) -> int:
    """
    Time Complexity O(N)
    Memory Complexity O(H)
    """
    max_val: int = float('-inf')
    cur_nodes: list[Node] = [root]
    next_nodes: list[Node] = []
    while cur_nodes:
        for node in cur_nodes:
            max_val = max(max_val, node.value)
            if node.left:
                next_nodes.append(node.left)
            if node.right:
                next_nodes.append(node.right)
        cur_nodes = next_nodes
        next_nodes = []
    return max_val


def test():
    node1 = Node(1)
    node2 = Node(-5)
    node3 = Node(3, node1, node2)
    node4 = Node(2, node3, None)
    assert solution(node4) == 3


if __name__ == '__main__':
    test()

File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/A._Лампочки/solution.cpp
#ifndef REMOTE_JUDGE
struct Node {
  int value;
  const Node* left = nullptr;
  const Node* right = nullptr;
  Node(int value, Node* left, Node* right) : value(value), left(left), right(right) {}
};
#endif

#ifdef REMOTE_JUDGE
#include "solution_tree.h"
#endif
#include <cmath>
#include <iostream>
#include <cassert>
#include <vector>
#include <cstdint>

using namespace std;


int Solution(const Node* root) {
    /*
    Time Complexity O(N)
    Memory Complexity O(H)
    */
    int maxVal = INT32_MIN;
    std::vector<const Node*> curNodes = {root};
    std::vector<const Node*> nextNodes;
    while (curNodes.size() > 0) {
        for (const Node* node : curNodes) {
            maxVal = std::max(maxVal, node->value);
            if (node->left) {
                nextNodes.push_back(node->left);
            }
            if (node->right) {
                nextNodes.push_back(node->right);
            }
        }
        curNodes = nextNodes;
        nextNodes = {};
    }
    return maxVal;
}

#ifndef REMOTE_JUDGE
void test() {
    Node node1({1, nullptr, nullptr});
    Node node2({-5, nullptr, nullptr});
    Node node3({3, &node1, &node2});
    Node node4({2, &node3, nullptr});
    assert(Solution(&node4) == 3);
}

int main() {
  test();
}
#endif